File: AtCoder Beginner Contest 304/Good Graph.py
import bisect
from collections import deque
import math
import sys
from typing import List
from heapq import *


def LI(): return list(map(int, sys.stdin.readline().rstrip().split()))


def II(): return int(sys.stdin.readline())


def MI(): return map(int, sys.stdin.readline().rstrip().split())


def S(): return sys.stdin.readline().rstrip()


def SI(): return sys.stdin.readline().strip()


def MIN_INT(): return -1 * sys.maxsize


def MAX_INT(): return sys.maxsize


readline = sys.stdin.readline
sys.setrecursionlimit(10 ** 7)


class DSU:
    def __init__(self, n):
        self.parent = list(range(n + 1))
        self.depth = [1] * (n + 1)
        self.forests = n

    def find(self, x):
        if self.parent[x] == x:
            return x
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def link(self, x, y):
        if self.depth[x] > self.depth[y]:
            x, y = y, x
        self.parent[x] = y
        if self.depth[x] == self.depth[y]:
            self.depth[y] += 1

    def union(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x != y:
            self.link(x, y)
            self.forests -= 1
        return x != y

def solve():
    n, m = MI()
    d = DSU(n)
    for _ in range(m):
        u, v = MI()
        d.union(u, v)

    no = set()
    k = II()
    for _ in range(k):
        u, v = MI()
        x = d.find(u)
        y = d.find(v)
        if x > y:
            x, y = y, x
        no.add((x, y))

    q = II()
    for _ in range(q):
        u, v = MI()
        x = d.find(u)
        y = d.find(v)
        if x > y:
            x, y = y, x
        if (x, y) in no:
            print("No")
        else:
            print("Yes")


if __name__ == '__main__':
    t = 1  # II()
    for _ in range(t):
        solve()

File: AtCoder Beginner Contest 304/Good Graph.cpp
/*
 ####  ####  ####
 ## ##  ##  ## ##
 ##  ## ## ##  ##
 ##   ######   ##
 ##     ## ######
 ###   ####   ###
*/
#define _CRT_SECURE_NO_WARNINGS
#include<bits/stdc++.h>
#include<unordered_map>
#include<unordered_set>
#define MOD1 1000000007
#define MOD2 998244353
#define pi acos(-1)
#define ull unsigned long long
#define ll long long
#define all(s) s.begin(),s.end()
#define allr(s) s.rbegin(),s.rend()
#define OO 1000000000
#define mp(x,y) make_pair(x,y)
using namespace std;
long long Mod(long long num, long long mod) { return ((num % mod) + mod) % mod; }
long long Ceil(long long num, long long div) { return (num + div - 1) / div; }
long long Gcd(long long a, long long b) { if (a == 0)return b; return Gcd(b % a, a); }
long long Lcm(long long a, long long b) { return a / Gcd(a, b) * b; }
int quadform(long long n) { return (ceil(sqrt((8 * n)) + 1) - 1) / 2; }
const int dx[] = { -1,0,1,0,-1,1,-1,1 };
const int dy[] = { 0,1,0,-1,-1,1,1,-1 };

void fast()
{
    std::ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
#ifndef ONLINE_JUDGE
    freopen("in.txt", "r", stdin);
    freopen("out.txt", "w", stdout);
#endif
}
struct DSU {
	vector<int>parent, depth;
	int forests = 0;
	DSU(int n) {
		parent = depth = vector<int>(n + 1);
		for (int i = 1; i <= n; i++) {
			parent[i] = i;
			depth[i] = 1;
			forests++;
		}
	}
	int Find(int x) {
		if (parent[x] == x)
			return x;
		return parent[x] = Find(parent[x]);
	}
	void Link(int x, int y) {
		if (depth[x] > depth[y])
			swap(x, y);
		parent[x] = y;
		if (depth[x] == depth[y])
			depth[y]++;
	}
	bool Union(int x, int y) {
		x = Find(x), y = Find(y);
		if (x != y) {
			Link(x, y);
			forests--;
		}
		return x != y;
	}
};
void solve()
{
	int n, m;
	cin >> n >> m;
	DSU d(n);
	int u, v;
	for (int i = 0; i < m; i++) {
		cin >> u >> v;
		d.Union(u, v);
	}
	set<vector<int>>no;
	int k;
	cin >> k;
	int ok = 1;
	for (int i = 0; i < k; i++) {
		cin >> u >> v;
		int x = d.Find(u), y = d.Find(v);

		if (x > y)
			swap(x, y);
		no.insert({ x,y });

	}

	int q;
	cin >> q;
	for (int i = 0; i < q; i++) {
		cin >> u >> v;
		int x = d.Find(u), y = d.Find(v);
		if (x > y)
			swap(x, y);
		cout << (no.find({ x,y }) != no.end() ? "No" : "Yes") << "\n";
	}
}

int main() {
    fast();
    int t = 1;
    //cin >> t;
    for (int tt = 1; tt <= t; tt++) {
        //cout << "Case " << tt << ": ";
        solve();
        /* if (tt < t)
             cout << "\n";*/
    }

    return 0;
}
